### ISSUE TITLE
Make CKEDITOR a generic events proxy

### ISSUE THREAD

**Issue by fredck** [ckeditor/ckeditor5-design#53]
We've been talking about the possibility of having "event bubbling". For example having a chain of events fired on an object's parents tree, like: button.press (source) -> toolbar.press -> region.press -> editor.press. We understood that this is tricky.

In the other hand, it would be extremely useful to handle a single place where to listen for events from any source and of any kind. The optimal place for it would be the `CKEDITOR` module.

This means that anything that uses `fire()` from our events API would fire also on `CKEDITOR`. Then, the `source` property of the event object passed to listeners can be used to recognize the source.

Acceptable?

**Comment by mlewand** [comment]
Yes that would be very cool. It will require us to maintain a proper parent-child relation in objects.

E.g. the parent of button would be a toolbar, etc.

**Comment by Reinmar** [comment]
In my opinion it would be too much if everything was bubbled up to the `CKEDITOR` object. In many cases this would be totally unnecessary and it could be really bad for performance. I'm still for explicit forwarding - we can have some events forwarded to `CKEDITOR` object, but only if that makes sense.

**Comment by pjasiun** [comment]
I think it will be very useful for users to just listen on CKEDITOR and do not care about the instances.

In many cases user wants to configure some generic behavior: additional action on paste or standard server response parsing, and he wants to set it in the very different place in code that where the `editor` instance is created. So big +1.

**Comment by fredck** [comment]
> It will require us to maintain a proper parent-child relation in objects.

@mlewand, I'm not sure you understood the issue... it's about NOT going with event bubbling but instead have a single place where to have all events fired: the `CKEDITOR` module.

**Comment by pjasiun** [comment]
I think that performance is the matter of the implementation, it does not need to be an issue.

**Comment by fredck** [comment]
@Reinmar I don't want to complicate things... it should be a dumb approach... and because of this, it may  have irrelevant performance impact.

**Comment by mlewand** [comment]
> it's about NOT going with event bubbling

Ahh, my bad got it wrong.

So you're saying it's going to fire only in 2 objects, that's the source and the CKEDITOR ns object? Well then it's a little bit different storry.

**Comment by fredck** [comment]
> So you're saying it's going to fire only in 2 objects, that's the source and the CKEDITOR ns object?

Exactly.

**Comment by pjasiun** [comment]
And I think we should keep it simple. If all events are forwarded to CKEDITOR users will not need to check and if only part of then, then it could be a reason of bugs when user will believe that the event is forwarded and it is not, he need to check each event in the docs, and he need to know that not all of then are forwarded.

**Comment by oleq** [comment]
@fredck Correct me if I'm wrong: you want to listen on `CKEDITOR` to the `press` event fired on `button` but with `source: button`?

**Comment by Reinmar** [comment]
> @mlewand, I'm not sure you understood the issue... it's about NOT going with event bubbling but instead have a single place where to have all events fired: the CKEDITOR module.

I think that you guys missed this point. Fred's idea is to forward **every** event fired anywhere to `CKEDITOR`. So we are talking about hundreds of them. Every single component, every single data change, selection change, button state change... everything.

So, we'll have code like:

```
CKEDITOR.on( 'change:state', function( evt ) {
    if ( evt.source instanceof Button && evt.source.name == 'Bold' ) {
        // do something...
    } 
} );
```

How is this good comparing to?

```
editor.ui.get( 'Bold' ).on( 'change:state',  function() {
} );
```

Of course this is a stupid example. But that's why I wrote that it's unnecessary - because in 99% of cases there's a better way to do the same.

**Comment by adelura** [comment]
I'm for it, therefore we could implement it more generic way instead of hardcoding it into CKEDITOR:

``` js
this.fire( 'event', {
    someProperty: 'some value'
}, CKEDITOR  /* This means that proxy is set up on CKEDITOR and will be fired there as well. */ );
```
If no extra argument is provided, then no event fired on proxy.

**Comment by oleq** [comment]
@Reinmar :+1:

**Comment by pjasiun** [comment]
Can we agree that all `editor`'s events should be forwarded to CKEDITOR?

**Comment by fredck** [comment]
@Reinmar, again, this is a generic approach... you'll always have a way to listen directly over the source as well.

For instance... if you would like to listen for all `change:state` in all `Button`'s (stupid example as well), it would be much more complicated than what you exemplified.

**Comment by fredck** [comment]
Btw, currently `fire()` is pretty fast if there is no registered listeners... the first thing it does is a dictionary check for them and it exits straight if none found. We can even try to micro-optimize it if necessary.

I don't see this as a performance issue.

**Comment by mlewand** [comment]
> I think that you guys missed this point.

Yea, well then we'll end up with tons of events being propagated and __evaluated__.

E.g.

```javascript
CKEDITOR.on( 'key', function( evt ) {
	// Generic Ui button listener.
	if ( evt.source instanceof Button && evt.keyCode == keyCodes.SPACE ) {
		evt.source.click();
	}
} );

CKEDITOR.on( 'key', function( evt ) {
	// Listener for widgets.
	if ( helper.isWidget( evt.source ) && evt.keyCode == keyCodes.ENTER ) {
		evt.source.edit();
	}
} );
```

Events propagation is awesome, but I have not seen UI lib that fires events only for 2 objects. It's always bubbling through parent-child list.

**Comment by Reinmar** [comment]
This feature will be useful only when you want to listen to all `xyz` events regardless of its source. In all other cases you will need to check the source. For instance, if you want to listen to `change:state` on all buttons you will need to check whether source is a button and not something else.

For me it makes much more sense to forward `change:state` events to buttons collection. There it is useful.

**Comment by Reinmar** [comment]
But again - it makes sense to forward some **editor** events to `CKEDITOR` or `CKEDITOR.instances` (since it's a collection of editors). We can even forward all editor events to `CKEDITOR.instances` if you want. But not all events from every unrelated source.

**Comment by fredck** [comment]
Well, we're talking again about event bubbling... we discussed this already and I've been clearly agains it, being supported by some of you guys.

So if we're talking about bubbling again, then we're missing the point of the issue.

This is all about a **tool** to make it easier to have a "catch all" proxy for events. It is not the one and only way to listen to events. It is not the way we gonna show events in our documentation. 

> For instance, if you want to listen to change:state on all buttons you will need to check whether source is a button and not something else.

Yes, it is all about that. It is just a shortcut, with no parent/child relation, nothing else. Pure KISS.

**Comment by pjasiun** [comment]
> Events propagation is awesome, but I have not seen UI lib that fires events only for 2 objects. It's always bubbling through parent-child list.

I like the idea of bubbling. It would be very useful if I could listen on all widgets on WidgetsRepository, on all uploads on UploadRepositor or on all button on toolbar. It would make code much simpler.

**Comment by fredck** [comment]
So, if using `CKEDITOR` as a proxy is (for some reason still not clear) not acceptable, the one and only option I see for this is bubbling from `Model` to its relative `Collection`.

This has been in fact proposed in the first implementation of `Collection` but it has been removed with https://github.com/cksource/ckeditor5-core/commit/edc9dadea82d2e8e0652671cee8a11422f0d43d9. The discussion about it can be found here:
https://github.com/ckeditor/ckeditor5-core/pull/13#discussion_r24582699

**Comment by pjasiun** [comment]
The only problem with CKEDITOR as a global proxy I see is that all events names should be unique or sender should be checked every time. Otherwise if one plugin introduce some local events with the same name what other plugin, then user can meet unexpected behavior.

But I still do not see the problem with having too many events on `CKEDITOR`. Of course some of them should not be used and if user assign some heavy listeners on events which are fired very often in may cause performance issue, but this is up to the user how he use our API. We should make it as simple and elastic as possible.

**Comment by Reinmar** [comment]
> So, if using CKEDITOR as a proxy is (for some reason still not clear) not acceptable, the one and only option I see for this is bubbling from Model to its relative Collection.

Both options have -1 from me.

* I don't like automated bubbling because the semantics is broken.
* I don't like making `CKEDITOR` a proxy because the semantics is broken even more. Plus, usage of such events will be very hard for users because they will need to understand how to correctly check the source. Plus, bubbled events may have the same names as direct `CKEDITOR` events - it will be a disaster if `CKEDITOR` has any direct events and someone tries to use them... So I don't see any added value here. It slightly reduces one problem and creates bunch of new ones...

I would approach this from other side. Less automated, but much more clean, because precise and direct. You want an item event on a collection? Forward it. Forwarding can be automated to make it concise:

```
var Collection = Collection.extend( { 
    hasItemEvents: [
        'change:state item:change:state'
    ]
} );
```

Plus we can have additional methods like https://github.com/ckeditor/ckeditor5-design/issues/7#issuecomment-74501808 or listening on all collection items' events even when there's no forwarding.

In other words - I think that taking a dirty shortcuts just to solve some cases is wrong if you can solve more cases taking a little bit longer path. But of course making `CKEDITOR` a proxy is cheap to implement and cheap to remove if it'll turn out that there are problems, so go ahead if you want it.

**Comment by Reinmar** [comment]
BTW. Can this be implemented without making `CKEDITOR` a direct dependency of the event emitter? This kind of decisions will make the code less reusable. For instance, CKFinder will not be able to reuse CKEditor's emitter without having `CKEDITOR` object floating around. We will make every single class dependent on `CKEDITOR`.

**Comment by fredck** [comment]
> bubbled events may have the same names as direct CKEDITOR events

That's the exactly reason for me to -1 bubbling in the past and so, here again, it's a -1 for me as well. The idea of using CKEDITOR for proxing is not good.

I still like the idea of automatically firing Model methods into its relative Collection. After all collections are supposed to have the "add" and "remove" events only, so we could leave these names reserved for them, avoiding collision..

**Comment by Reinmar** [comment]
There's no `CKEDITOR` singleton so there's no proxy :)
